### DISCUSSION TITLE
What is the bare minimum version of React required?

### DISCUSSION THREAD

**Discussion by jgornick** [nativewind/nativewind#370] [General]
We've run into a few situations when using versions of NativeWind 2.0.10+ (including 3.x) where because the `tsconfig.json` `jsx` value changed to [`react-jsx`, it requires a version of React 17+](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#react-17-jsx-factories).

This means the distribution of NativeWind is requiring React 17+, but it's not defined as a peer dependency.

My question is does NativeWind utilize features in React 17+? If not, what is the minimum version of React required? If so, should it be added as a peer dependency?

For what it's worth, a project I'm on is still using React 16.

**Comment by marklawlor** [comment]
I haven't put to much thought into it, but I'll aim to make this the policy and add them peerDependencies 

- NodeJS: Current & LTS version and the latest patch version of maintenance
- Current major version of TailwindCSS (new majors will be a major NativeWind version)

For React, I aiming for 2 years of support, which would be
- React 17.0
- React Native 0.63

Which are in conflict of each other because the RN team were slow to update to React 17 and 0.64 didn't come out until March 2021. As v3 won't be released until 2023 and 0.63 will be close to the 2 year cutoff,  I think its ok to make the minimum 0.64.

As the project matures and has other contributors to help maintain multiple versions I'll increase the support window, but for now I'm going to it keep reasonability tight.

NativeWind uses the features of React 18 which fortunately can be shimmed for previous versions of React.

**Reply by jgornick** [reply]
Thank you for the detailed response @marklawlor!

> NativeWind uses the features of React 18 which fortunately can be shimmed for previous versions of React.

To save me from digging a little bit, is it possible off the top of your head to provide which features and/or what shims? If it's a bit more than just a quick response, I can do some digging 😄 

Thanks again for all the awesome work!

**Reply by marklawlor** [reply]
> To save me from digging a little bit, is it possible off the top of your head to provide which features and/or what shims?

https://beta.reactjs.org/apis/react/useSyncExternalStore
https://www.npmjs.com/package/use-sync-external-store

**Reply by marklawlor** [reply]
And I also have upcoming plans to utilise `useId()`

**Comment by danstepanov** [comment]
Closing this discussion, if this is still relevant, feel free to open a new one.
